common/src/main/scala/uk/ac/wellcome/s3/S3ObjectStore.scala

```diff
@@ -0,0 +1,100 @@
+package uk.ac.wellcome.s3
+
+import java.security.MessageDigest
+
+import com.amazonaws.services.s3.AmazonS3
+import com.google.inject.Inject
+import com.twitter.inject.Logging
+import com.twitter.inject.annotations.Flag
+import io.circe.{Decoder, Encoder}
+import uk.ac.wellcome.models.Sourced
+import uk.ac.wellcome.utils.JsonUtil
+import uk.ac.wellcome.utils.GlobalExecutionContext.context
+
+import scala.concurrent.Future
+import scala.io.Source
+import scala.util.hashing.MurmurHash3
+
+// KeyPrefixGenerator is contravariant for type T
+//
+// Practically this means `KeyPrefixGenerator[Sourced]` is a
+// subclass of `KeyPrefixGenerator[SierraTransformable]`
+// so it will be accepted in the S3ObjectStore for all types
+// extending Sourced. This prevents us from having to write
+// SierraTransformableKeyPrefixGenerator etc. etc.
```

Comment:
Yay comments! Does this still make sense to you today?

I might go further, and be explicit about the relationship between `Sourced` and `SierraTransformable` (which may not be obvious if you're not familiar with our codebase.

Something like (spoilered if you'd rather have a go at writing it yourself):

<details>
<summary>Spoiler block</summary>

```scala
// '-T' means KeyPrefixGenerator is contravariant for type T
//
// This means that if S is a subclass of T, then KeyPrefixGenerator[T]
// is a subclass of KeyPrefixGenerator[S] (the class hierarchy is reversed).
//
// For example,
//
//      Sourced < SierraTransformable
//
// but
//
//      KeyPrefixGenerator[SierraTransformable] < KeyPrefixGenerator[Sourced]
//
// In S3ObjectStore, we need a KeyPrefixGenerator[T], where T is usually some
// subclass of Sourced (this is the way we use it, not a strict requirement).
// This contravariance allows us to define a single KeyPrefixGenerator[Sourced]
// and use it on all instances, even though T is some subclass of Sourced.
```

</details>

Reply:
Still makes sense, but i'm going to steal your spoiler!